inference-engine/src/mkldnn_plugin/nodes/mkldnn_input_node.cpp

```diff
@@ -3,27 +3,280 @@
+
+namespace {
+
+struct jit_has_subnormals : public jit_generator {
+    DECLARE_CPU_JIT_AUX_FUNCTIONS(jit_has_subnormals)
+
+    typedef struct {
+        const float* src;
+        const size_t count;
+        bool hasSubnormals;
+    } args_t;
+
+    typedef void (*fn_t)(const args_t*);
+
+    jit_has_subnormals() {
+        jit_ker_ = nullptr;
+    }
+
+    fn_t get() {
+        return jit_ker() || create_kernel() == dnnl::impl::status::success
+                ? (fn_t)jit_ker()
+                : nullptr;
+    }
+
+protected:
+    void foreach(const Xbyak::Reg64& idx,
+                 size_t step,
+                 const Xbyak::Reg64& end,
+                 std::function<void(const Xbyak::Reg64&)> && fn) {
+        Label loop, exit;
+
+        L(loop);
+        cmp(idx, end);
+        jge(exit);
+
+        fn(idx);
+
+        add(idx, step);
+        jmp(loop);
+        L(exit);
+    }
+
+    void copy_floats(const Xbyak::Reg64& dst,
+                     const Xbyak::Reg64& src,
+                     const Xbyak::Reg64& size) {
+        push(rsi);
+        push(r15);
+
+        xor_(rsi, rsi);
+
+        foreach(rsi, 1, size, [&, this](const Xbyak::Reg64& idx) {
+            mov(r15d, dword[src + idx * sizeof(float)]);
+            mov(dword[dst + idx * sizeof(float)], r15d);
+        });
+
+        pop(r15);
+        pop(rsi);
+    }
+};
+
+struct jit_has_subnormals_avx2 : public jit_has_subnormals {
+    static const uint32_t vlen = 8u;    // floats vector length for AVX2 instructions
+
+    void check_subnormals(const Xbyak::Reg64& src, const Xbyak::Ymm &mask, const Xbyak::Ymm &zero) {
+        auto a = ymm1;
+        auto b = ymm2;
+        auto c = ymm3;
+
+        vmovdqu(a, yword[src]);         // load 8 floats
+        vpcmpeqd(b, a, zero);           // if (a == 0) b = 1 else b = 0
+        vpand(c, a, mask);              // c = a & 01111111100000000000000000000000
+        vpcmpeqd(c, c, zero);           // if (c == 0) c = 1 else c = 0
+        vptest(b, c);                   // if ((!b & c) == 0) CF = 1 else CF = 0
+    }
+
+    void generate() final {
+        Label exit, has_subnormals, no_subnormals;
+
+        auto reg_src = rax;
+        auto reg_dst = rbx;
+        auto reg_sz = rdx;
+        auto reg_mask_addr = r15;
+        auto zero = ymm4;
+        auto mask = ymm5;
+
+        preamble();
+
+        // Initialize necessary consts
+        vpxor(zero, zero, zero);
+
+        static const uint32_t mask_data[8] = {
+            0xFF << 23, 0xFF << 23, 0xFF << 23, 0xFF << 23,
+            0xFF << 23, 0xFF << 23, 0xFF << 23, 0xFF << 23
+        };
+
+        mov(reg_mask_addr, (size_t)mask_data);
+        vmovdqu(mask, yword[reg_mask_addr]);
+
+        // Get arguments addresses
+        mov(reg_src, ptr[param1 + offsetof(args_t, src)]);
+        lea(reg_dst, ptr[param1 + offsetof(args_t, hasSubnormals)]);
+        mov(reg_sz, ptr[param1 + offsetof(args_t, count)]);
+
+        // Main loop
+        xor_(rsi, rsi);
```

Comment:
does it make sense to get rsi meaningful name as for other registers?

Reply:
fixed